### PR TITLE
[coverage] fix crash in code coverage and `if constexpr` with `ExprWithCleanups`

### LINKED ISSUE #80285: [coverage] crash with ExprWithCleanups in `if constexpr`
Clang `main` and `18` crashes with this code: (`-fprofile-instr-generate -fcoverage-mapping -std=c++17`)

```c++
struct false_value {
	constexpr operator bool() {
		return false;
	}
};

void foo()
{
	if constexpr (false_value{}) { };
}
```

https://godbolt.org/z/833ej19rP

It's a bug in my recent change https://github.com/llvm/llvm-project/pull/78033 ... writing patch now, this is so it's visible.

### PR DESCRIPTION
**Pull request by hanickadot**
Fixes https://github.com/llvm/llvm-project/issues/80285

### CODE REVIEW

File: clang/lib/CodeGen/CoverageMappingGen.cpp [hunk 1]
```diff
@@ -1808,12 +1808,24 @@ struct CounterCoverageMappingBuilder
...
+      return Expr->getResultAsAPSInt().getExtValue();
+
+    if (const auto *Expr = dyn_cast<ExprWithCleanups>(Condition))
+      return evaluateConstantCondition(Expr->getSubExpr()); // recursion
+
+    assert(false && "Unexpected node in 'if constexpr' condition");
+    return false;
+  }
+
+public:
```
Comment (cor3ntin) [inline clang/lib/CodeGen/CoverageMappingGen.cpp:1823]: @AaronBallman This is basically an `IgnoreImpCast` except we want to stop at the first ConstantExpr (which is a FullExpr) - do you think this is worth adding a way to do that In `Expr`?

Comment (efriedma-quic) [reply clang/lib/CodeGen/CoverageMappingGen.cpp:1823]: Can we just use IfStmt::getNondiscardedCase here?

More generally, we could add a method for evaluating ConstantExprs, I guess, but IfStmt itself currently just uses EvaluateKnownConstInt, which should be equivalent for the relevant cases.

Comment (cor3ntin) [reply clang/lib/CodeGen/CoverageMappingGen.cpp:1823]: I missed this was a thing. Yes, we should use that!

Comment (hanickadot) [reply clang/lib/CodeGen/CoverageMappingGen.cpp:1823]: I'm still a bit new here and I'm confused... `CoverageMappingGen.cpp` doesn't have any mention of ASTContext which is an argument for functions mentioned by @efriedma-quic is there an easy way how to get it. Or do I need to go another direction?

Comment (cor3ntin) [reply clang/lib/CodeGen/CoverageMappingGen.cpp:1823]: `CVM.getCodeGenModule().getContext()` from within `CounterCoverageMappingBuilder` might work

Comment (hanickadot) [reply clang/lib/CodeGen/CoverageMappingGen.cpp]: Ok it seems to work :)

Comment (hanickadot) [reply clang/lib/CodeGen/CoverageMappingGen.cpp]: Thanks!

**Review by AaronBallman — COMMENTED** [review]
Changes LGTM, but please add a release note to `clang/docs/ReleaseNotes.rst` so users know about the fix.

Thank you for the fix!

**Comment by hanickadot** [comment]
> Changes LGTM, but please add a release note to `clang/docs/ReleaseNotes.rst` so users know about the fix.
> 
> Thank you for the fix!

Even if it's a fix for recent PR I did few days ago?

Also, how do I get this into release branch (the previous change is there)?

File: clang/test/CoverageMapping/if.cpp [hunk 2]
```diff
@@ -234,6 +234,35 @@ constexpr int check_macro_consteval_if_skipped(int i) {   // CHECK-NEXT: [[@LINE
...
+
+template <typename> struct dependable_false_value {
+  constexpr operator bool() {
+    return false;
+  }
+};
+
+// GH-80285
+void should_not_crash() {
+    if constexpr (false_value{}) { };
```
Comment (MaskRay) [inline clang/test/CoverageMapping/if.cpp:251]: indent by 2

**Review by MaskRay — APPROVED** [review]

File: clang/test/CoverageMapping/if.cpp (hunk 2 above)
Comment (hanickadot) [reply clang/test/CoverageMapping/if.cpp]: Thanks!

**Comment by cor3ntin** [comment]
> Changes LGTM, but please add a release note to `clang/docs/ReleaseNotes.rst` so users know about the fix.
> 
> Thank you for the fix!

FYI this is a regression in 18 affecting users so we should backport

**Comment by AaronBallman** [comment]
> > Changes LGTM, but please add a release note to `clang/docs/ReleaseNotes.rst` so users know about the fix.
> > Thank you for the fix!
> 
> FYI this is a regression in 18 affecting users so we should backport

Btw, instructions on how to do that can be found here: https://www.llvm.org/docs/GitHub.html#backporting-fixes-to-the-release-branches

**Review by cor3ntin — APPROVED** [review]

**Pull request merged**